File: NovaSim/NovaSim.Infrastructure/Events/EventBus.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using NovaSim.Core.Interfaces;

namespace NovaSim.Infrastructure.Events
{
    public class EventBus : IEventBus
    {
        private static readonly Lazy<EventBus> _instance = new(() => new EventBus());
        private readonly ConcurrentDictionary<string, List<(Action<EventPayload> callback, string id)>> _listeners = new();
        private readonly List<(Action<string, EventPayload> callback, string id)> _wildcardListeners = new();
        private readonly object _lock = new();

        private EventBus() { }

        public static EventBus Instance => _instance.Value;

        public Action Subscribe(string @event, Action<EventPayload> callback)
        {
            var id = Guid.NewGuid().ToString("N");
            lock (_lock)
            {
                var list = _listeners.GetOrAdd(@event, _ => new List<(Action<EventPayload>, string)>());
                list.Add((callback, id));
            }

            return () =>
            {
                lock (_lock)
                {
                    if (_listeners.TryGetValue(@event, out var list))
                    {
                        list.RemoveAll(item => item.id == id);
                    }
                }
            };
        }

        public Action SubscribeAll(Action<string, EventPayload> callback)
        {
            var id = Guid.NewGuid().ToString("N");
            lock (_lock)
            {
                _wildcardListeners.Add((callback, id));
            }

            return () =>
            {
                lock (_lock)
                {
                    _wildcardListeners.RemoveAll(item => item.id == id);
                }
            };
        }

        public void Publish(string @event, string sender, object? data = null)
        {
            var payload = new EventPayload(DateTime.UtcNow, sender, data);

            // Fetch exact matches
            List<(Action<EventPayload> callback, string id)>? exact = null;
            lock (_lock)
            {
                if (_listeners.TryGetValue(@event, out var list))
                {
                    exact = new List<(Action<EventPayload> callback, string id)>(list);
                }
            }

            if (exact != null)
            {
                foreach (var (callback, _) in exact)
                {
                    try
                    {
                        callback(payload);
                    }
                    catch (Exception ex)
                    {
                        Console.WriteLine($"[EventBus] Error in exact event handler for '{@event}': {ex.Message}");
                    }
                }
            }

            // Fetch wildcards
            List<(Action<string, EventPayload> callback, string id)> wildcards;
            lock (_lock)
            {
                wildcards = new List<(Action<string, EventPayload> callback, string id)>(_wildcardListeners);
            }

            foreach (var (callback, _) in wildcards)
            {
                try
                {
                    callback(@event, payload);
                }
                catch (Exception ex)
                {
                    Console.WriteLine($"[EventBus] Error in wildcard event handler for '{@event}': {ex.Message}");
                }
            }
        }

        public void ClearAllListeners()
        {
            lock (_lock)
            {
                _listeners.Clear();
                _wildcardListeners.Clear();
            }
        }
    }
}

File: src/infrastructure/services/EventBus.ts
import { IEventBus } from "../../domain/services/IEventBus";
import { EngineEvents, EventPayload, EventCallback } from "../../shared/events/EngineEvents";

export class EventBus implements IEventBus {
  private static instance: EventBus | null = null;
  private listeners: Map<string, Array<{ callback: EventCallback; id: string }>> = new Map();
  private wildcardListeners: Array<{ callback: (event: string, payload: EventPayload) => void; id: string }> = [];

  private constructor() {}

  public static getInstance(): EventBus {
    if (!EventBus.instance) {
      EventBus.instance = new EventBus();
    }
    return EventBus.instance;
  }

  public subscribe(event: EngineEvents | string, callback: EventCallback): () => void {
    const id = Math.random().toString(36).substring(2, 11);
    
    if (!this.listeners.has(event)) {
      this.listeners.set(event, []);
    }
    
    this.listeners.get(event)!.push({ callback, id });

    return () => {
      const list = this.listeners.get(event);
      if (list) {
        this.listeners.set(
          event,
          list.filter((l) => l.id !== id)
        );
      }
    };
  }

  public subscribeAll(callback: (event: string, payload: EventPayload) => void): () => void {
    const id = Math.random().toString(36).substring(2, 11);
    this.wildcardListeners.push({ callback, id });

    return () => {
      this.wildcardListeners = this.wildcardListeners.filter((w) => w.id !== id);
    };
  }

  public publish(event: EngineEvents | string, sender: string, data?: any): void {
    const payload: EventPayload = {
      timestamp: new Date(),
      sender,
      data
    };

    // Dispatch exact match
    const exactListeners = this.listeners.get(event);
    if (exactListeners) {
      for (const listener of exactListeners) {
        try {
          listener.callback(payload);
        } catch (err) {
          console.error(`[EventBus] Error in listener for event ${event} triggered by ${sender}:`, err);
        }
      }
    }

    // Dispatch wildcards
    for (const wild of this.wildcardListeners) {
      try {
        wild.callback(event, payload);
      } catch (err) {
        console.error(`[EventBus] Error in wildcard subscriber for event ${event}:`, err);
      }
    }
  }

  public clearAllListeners(): void {
    this.listeners.clear();
    this.wildcardListeners = [];
  }
}
